### PR TITLE
[JENKINS-32155] Recognize new repositories automatically

### PR DESCRIPTION
**Pull request by recena**
[JENKINS-32155] (https://issues.jenkins-ci.org/browse/JENKINS-32155)

- [x] Webhook autoregistration for organizations and repositories
- [x] New repositories are processed automatically

@reviewbybees

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -33,6 +33,11 @@
         </dependency>
         <dependency>
             <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>branch-api</artifactId>
```
Comment (jglick) [inline pom.xml]: :bug: for this dependency. If `scm-api` is missing some APIs we need for this feature, they should be added there, in a style consistent with the abstracted design of `SCMSourceOwners` etc., with any required implementations added to `branch-api`.

Comment (recena) [inline pom.xml]: I totally disagree. I don't see the worth of overloading `scm-api`.

Comment (recena) [inline pom.xml]: It is the second PR (from two different developers) that you rejected by the same reason. Maybe something is wrong here.

**Comment by recena** [comment]
I'm playing with this PR and it seems work fine.

**Comment by jglick** [comment]
@stephenc may wish to advise.

**Comment by amuniz** [comment]
I think the designed architectural restriction applies - and makes sense - for `SCMSource`/`SCMNavigaror` implementations, but in this case `OrganizationFolder` is being used to distinguish where to auto-create a webhook, which is totally decoupled from the `SCM*` implementations, so architecturally speaking it's good to me.

OTOH adding that dependency opens the Pandora's box, and someone could be tempted to use `branch-api` stuff in the `SCM*` implementations in the future, which is clearly wrong.

My +1 is for adding a new entity in `scm-api` for `SCMNavigator` owners (`SCMNavigatorOwner`) as it is currently for SCMSource owners (`SCMSourceOwner`) - so `OrganizationFolder` implements it.

**Comment by recena** [comment]
@amuniz This PoV makes me to think about if `scm-api` and `branch-api` should not be the same dependency.

**Comment by recena** [comment]
@jglick I've started a new approach based on to avoid adding `branch-api` dependency /cc @stephenc

File: src/main/java/org/jenkinsci/plugins/github_branch_source/WebhookRegisterListener.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            return;
+        }
+    }
+
+    @Override
+    public void onUpdated(Item item) {
+        if (!isApplicable(item)) {
+            return;
+        }
+        if (item instanceof MultiBranchProject) {
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/WebhookRegisterListener.java]: Bad code smell here. I think this approach will result in a hacky API. much better to make the appropriate API exposure in scm-api

**Comment by stephenc** [comment]
I concur with @jglick's :bug:

**Comment by recena** [comment]
@stephenc I figured you'd say that. you are the father of these APIs ;) /cc @jglick

**Comment by stephenc** [comment]
@recena well there is a design and separation of concerns that backs why these two APIs are separate.

When you learn the appropriate way to expose features through these APIs you will understand why it has to be this way

Note: I said "appropriate" not "right/correct" as there may be better ways to design an SCM API **from scratch** and such ways might not mandate the separation of concerns that we have here... but the separation of concerns is very necessary for the code base we are integrating into

**Comment by recena** [comment]
This PR has been included in **GitHub Organization Folder** [1.0](https://github.com/jenkinsci/github-organization-folder-plugin/tree/github-organization-folder-1.0) by @kohsuke

**Comment by jglick** [comment]
Which I specifically objected to.

**Comment by recena** [comment]
I know it but I added the above comment was mainly to track the result of this PR and its Jira issue associated.

**Pull request closed without merging**